### PR TITLE
Refactor SuperblockValidator for by-project part combinations

### PR DESCRIPTION
**Pull request by cyrossignol**
Update the `SuperblockValidator` by-project fallback validation routine to check each combination of resolved manifest project parts that match the hints in the superblock.

The implementation of these changes should match the updated behavior of the reference `ValidateSuperblock()` function in the scraper.

### CODE REVIEW

File: src/neuralnet/superblock.cpp [hunk 1]
```diff
@@ -177,38 +208,299 @@ class SuperblockValidator
...
+    //! a hint can qualify more than one manifest part for consideration while
+    //! reconstructing a convergence for validation.
+    //!
+    //! This class behaves like a cursor through each combination of the parts
+    //! it is initialized with. For each call to \c GetNextConvergence(), this
+    //! class generates a convergence using a new combination of project parts
+    //! until it exhausts every combination or until the client code matches a
+    //! superblock with one of the results.
+    //!
+    //! We have built here a catapult that fires a huge boulder to kill a tiny
```
Comment (jamescowens) [inline src/neuralnet/superblock.cpp]: Haha. That is the truth! :)

File: src/neuralnet/superblock.cpp [hunk 2]
```diff
@@ -177,38 +208,299 @@ class SuperblockValidator
...
+
+        //!
+        //! \brief Generate the next convergence in sequence from the resolved
+        //! project parts.
+        //!
+        //! Jim Owens enlightens us with a stimulating description of his algorithm:
+        //!
+        //! Compute number of permutations and cumulative placevalue. For instance, if there were 4 projects, and
+        //! the number of candidate parts for each of the projects were 2, 3, 2, and 2, then the total number of
+        //! permutations would be 2 * 3 * 2 * 1 = 12, and the cumulative placevalue would be
```
Comment (jamescowens) [inline src/neuralnet/superblock.cpp]: I did a typo here in my original comments. I changed it midstream to be 2, 3, 2, 2 and forgot to update that part... it should be 2 * 3 * 2 * 2 = 24

Comment (cyrossignol) [reply src/neuralnet/superblock.cpp]: Fixed, along with a couple of other minor doc comments and tweaks 🙂

Comment (jamescowens) [reply src/neuralnet/superblock.cpp]: Merging!

**Pull request merged**